src/execution/compiler/operator/index_create_translator.cpp

```diff
@@ -0,0 +1,165 @@
+}
+
+void CreateIndexTranslator::DeclareSlot(FunctionBuilder *builder) {
+  // Get var slot = @pciGetSlot(pci)
+  ast::Expr *get_slot_call = codegen_->OneArgCall(ast::Builtin::PCIGetSlot, pci_, false);
+  builder->Append(codegen_->DeclareVariable(slot_, nullptr, get_slot_call));
+}
+
+void CreateIndexTranslator::GenPCILoop(FunctionBuilder *builder) {
+  // Generate for(; @pciHasNext(pci); @pciAdvance(pci)) {...} or the Filtered version
+  // The HasNext call
+  ast::Expr *has_next_call = codegen_->OneArgCall(ast::Builtin::PCIHasNext, pci_, false);
+  // The Advance call
+  ast::Expr *advance_call = codegen_->OneArgCall(ast::Builtin::PCIAdvance, pci_, false);
+  ast::Stmt *loop_advance = codegen_->MakeStmt(advance_call);
+  // Make the for loop.
+  builder->StartForStmt(nullptr, has_next_call, loop_advance);
+}
+
+void CreateIndexTranslator::GenCreateIndex(FunctionBuilder *builder) {
+  index_oid_ = codegen_->Accessor()->CreateIndexWrapper(op_->GetNamespaceOid(), op_->GetTableOid(), op_->GetIndexName(),
+                                                        *(op_->GetSchema()));
+}
+
+void CreateIndexTranslator::GenIndexInsert(FunctionBuilder *builder) {
+  std::vector<ast::Expr *> insert_args{codegen_->PointerTo(index_inserter_), codegen_->IntLiteral(!index_oid_),
+                                       codegen_->PointerTo(slot_)};
+  auto index_insert_call = codegen_->BuiltinCall(ast::Builtin::IndexInsertBulk, std::move(insert_args));
+  auto cond = codegen_->UnaryOp(parsing::Token::Type::BANG, index_insert_call);
+  builder->StartIfStmt(cond);
+  Abort(builder);
+  builder->FinishBlockStmt();
+}
+
+void CreateIndexTranslator::GenTVIClose(FunctionBuilder *builder) {
+  // Close iterator
+  ast::Expr *close_call = codegen_->OneArgCall(ast::Builtin::TableIterClose, tvi_, true);
+  builder->Append(codegen_->MakeStmt(close_call));
+}
+
+void CreateIndexTranslator::GenSetIndexLive(FunctionBuilder *builder) {
+  codegen_->Accessor()->SetIndexLive(index_oid_);
+}
+
+void CreateIndexTranslator::GenIndexInserterFree(FunctionBuilder *builder) {
+  // Call @storageInterfaceFree
+  ast::Expr *index_inserter_free = codegen_->OneArgCall(ast::Builtin::StorageInterfaceFree, index_inserter_, true);
+  builder->Append(codegen_->MakeStmt(index_inserter_free));
+}
+
+void CreateIndexTranslator::SetOids(FunctionBuilder *builder) {
+  // Declare: var col_oids: [num_cols]uint32
+  ast::Expr *arr_type = codegen_->ArrayType(all_oids_.size(), ast::BuiltinType::Kind::Uint32);
+  builder->Append(codegen_->DeclareVariable(col_oids_, arr_type, nullptr));
+
+  // For each oid, set col_oids[i] = col_oid
+  for (uint16_t i = 0; i < all_oids_.size(); i++) {
+    ast::Expr *lhs = codegen_->ArrayAccess(col_oids_, i);
+    ast::Expr *rhs = codegen_->IntLiteral(!all_oids_[i]);
+    builder->Append(codegen_->Assign(lhs, rhs));
+  }
+}
+
+std::vector<catalog::col_oid_t> CreateIndexTranslator::AllColOids(const catalog::Schema &table_schema_) {
```

Comment:
This seems like a utility function for `IndexSchema` if this is necessary.

Reply:
This function is used in other translators such as insert_translator, I don't know if we need make those translators all depend on IndexSchema

> This seems like a utility function for `IndexSchema` if this is necessary.

Reply:
Even worse. We shouldn't have the same function duplicated in 3 places in the code base. Suggest `storage_util.h/.cpp` if we can't find a more appropriate home.

Reply:
Moved to storge_util.h

> Even worse. We shouldn't have the same function duplicated in 3 places in the code base. Suggest `storage_util.h/.cpp` if we can't find a more appropriate home.